bee-crypto/src/ternary/bigint/u384/mod.rs

```diff
@@ -0,0 +1,482 @@
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+//! This module contains unsigned integers encoded by 384 bits.
+
+mod constants;
+
+pub use constants::{
+    BE_U32_0, BE_U32_1, BE_U32_2, BE_U32_HALF_MAX, BE_U32_HALF_MAX_T242, BE_U32_MAX, BE_U8_0, BE_U8_1, BE_U8_2,
+    BE_U8_MAX, LE_U32_0, LE_U32_1, LE_U32_2, LE_U32_HALF_MAX, LE_U32_HALF_MAX_T242, LE_U32_MAX, LE_U32_MAX_T242,
+    LE_U32_NEG_HALF_MAX_T242, LE_U32_ONLY_T243_OCCUPIED, LE_U8_0, LE_U8_1, LE_U8_2, LE_U8_MAX,
+};
+
+use crate::ternary::bigint::{
+    binary_representation::{BinaryRepresentation, U32Repr, U8Repr},
+    endianness::{BigEndian, LittleEndian},
+    error::Error,
+    overflowing_add::OverflowingAdd,
+    split_integer::SplitInteger,
+    t243, I384, T242, T243,
+};
+
+use bee_ternary::Utrit;
+
+use byteorder::{self, ByteOrder};
+
+use std::{
+    cmp::Ordering,
+    convert::TryFrom,
+    fmt,
+    marker::PhantomData,
+    ops::{Deref, DerefMut},
+};
+
+/// A big integer encoding an unsigned integer with 384 bits.
+///
+/// `T` is usually taken as a `[u32; 12]` or `[u8; 48]`.
+///
+/// `E` refers to the endianness of the digits in `T`. This means that in the case of `[u32; 12]`, if `E == BigEndian`,
+/// that the u32 at position i=0 is considered the most significant digit. The endianness `E` here makes no statement
+/// about the endianness of each single digit within itself (this then is dependent on the endianness of the platform
+/// this code is run on).
+///
+/// For `E == LittleEndian` the digit at the last position is considered to be the most significant.
+#[derive(Clone, Copy)]
+pub struct U384<E, T> {
+    pub(crate) inner: T,
+    _phantom: PhantomData<E>,
+}
+
+impl<E, T> Deref for U384<E, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<E, T> DerefMut for U384<E, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+impl<E: fmt::Debug, R: BinaryRepresentation, D> fmt::Debug for U384<E, R>
+where
+    E: fmt::Debug,
+    R: BinaryRepresentation<Inner = D>,
+    D: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("U384")
+            .field("inner", &self.inner.iter())
+            .field("_phantom", &self._phantom)
+            .finish()
+    }
+}
+
+impl U384<BigEndian, U32Repr> {
+    /// Reinterprets the U384 as an I384.
+    pub fn as_i384(self) -> I384<BigEndian, U32Repr> {
+        I384::<BigEndian, U32Repr>::from_array(self.inner)
+    }
+
+    /// Shifts the U384 in signed space.
+    pub fn shift_into_i384(mut self) -> I384<BigEndian, U32Repr> {
+        self.sub_inplace(*BE_U32_HALF_MAX);
+        self.sub_inplace(Self::one());
+        self.as_i384()
+    }
+
+    /// Adds `other` onto `self` in place.
```

Comment:
not sure about "onto". I think I'ld just write "to".

Reply:
It looks like it's mutating `self` though so 'onto' sounds like an appropriate word.

Reply:
alright. I won't argue with you Josh :smile: